### PR TITLE
Add data only stanza

### PR DESCRIPTION
**Pull request by rgrinberg**
subdirs and ignored_subdirs will now truly ignore directories. data-only can be
used to just ignore the rules

### CODE REVIEW

**Comment by rgrinberg** [comment]
This is now ready for review

**Comment by rgrinberg** [comment]
One last thing:

Do we want to maintain the old meaning of `ignored_subdirs`? I.e. have it add directories as `data_only`?

File: src/file_tree.ml [hunk 1]
```diff
@@ -70,48 +64,52 @@ module Dune_file = struct
...
+              | Sub_dirs x, None, _, [] -> (Some x, [], data_only)
+              | Sub_dirs _, Some _, _, _ ->
+                Errors.fail loc "More than one sub_dirs stanza is not allowed"
+              | Ignore_sub_dirs _, Some _, _, _
+              | Sub_dirs _, _, _, _::_ ->
+                Errors.fail loc
+                  "Cannot have both sub_dirs and ignored_sub_dirs \
+                   stanza in a dune file. "
+              | Data_only _, _, Some _, _ ->
+                Errors.fail loc "More than one data_only stanza is not allowed"
```
Comment (diml) [inline src/file_tree.ml]: A slightly nicer way to write this code would be something like this:

```ocaml
fields (
  let%map ignored_sub_dirs = multi_field "ignored_subdirs" ...
  and sub_dirs = field "sub_dirs" ...
  and data_only = field "data_only"
  and rest = leftover_fields
  in
 ...
)
```

where `leftover_field` is a new combinator to add that would return the unparsed fields as a `Sexp.Ast.t list`.

**Comment by diml** [comment]
To make things completely explicit, what about using `data_only_subdirs` rather than `data_only`?

**Comment by diml** [comment]
> Do we want to maintain the old meaning of ignored_subdirs? I.e. have it add directories as data_only?

Yes, we need to in order to preserve backward compatibility. In fact, let's deprecate `ignored_subdirs` now, it's meaning was always confusing.

File: src/file_tree.ml (hunk 1 above)
Comment (rgrinberg) [reply src/file_tree.ml]: I tried doing this but i can't get it to work. The last commit is WIP, will try again in a bit.

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/dune_lang/dune_lang.ml [hunk 2]
```diff
@@ -848,6 +848,14 @@ module Decoder = struct
     let x = result ctx (t ctx { unparsed; known = [] }) in
     (x, [])
 
+  let leftover_fields (Fields (_, _, _)) state =
+    ( Name_map.values state.unparsed
+      |> List.concat_map ~f:(fun x -> x.values)
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/dune_lang/dune_lang.ml]: This is wrong anyway btw, somehow some entries are being skipped. E.g. entries like:

```
(rule
 (with-stdout-to dune-config.5
  (run %{bin:jbuilder} help config --man-format=groff)))

```

File: src/sub_dirs.ml [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                (sub_dirs, i :: ignored, data_only)
+              | Data_only x, _, None, _ -> (sub_dirs, ignored, Some x)
+              | Sub_dirs x, None, _, [] -> (Some x, [], data_only)
+              | Sub_dirs _, Some _, _, _ ->
+                Errors.fail loc "More than one sub_dirs stanza is not allowed"
+              | Ignore_sub_dirs _, Some _, _, _
+              | Sub_dirs _, _, _, _::_ ->
+                Errors.fail loc
+                  "Cannot have both sub_dirs and ignored_sub_dirs \
+                   stanza in a dune file. "
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/sub_dirs.ml]: I think we should forbid both combinations. The combined behavior isn't really documented so I think its easier to just force people to only use the new stanza

File: src/dune_lang/dune_lang.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/dune_lang/dune_lang.ml]: Ah ok I fixed it. I was missing `prev`.

File: src/file_tree.ml [hunk 4]
```diff
@@ -234,88 +161,91 @@ let load ?(extra_ignored_subtrees=Path.Set.empty) path =
...
-              || Path.Set.mem extra_ignored_subtrees path
+                match (Sub_dirs.status sub_dirs ~dir:fn : Sub_dirs.Status.t)
+                with
+                | Normal -> (true, data_only)
+                | Ignored -> (false, data_only)
+                | Data_only -> (true, true)
             in
-            String.Map.add acc fn
-              (walk path ~dirs_visited ~project ~ignored))
+            if not is_subdir then
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/file_tree.ml]: Good point. Fixed.

**Comment by rgrinberg** [comment]
@diml ready for another round of review

File: src/sub_dirs.ml [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      | "" -> false
+      | s -> s.[0] <> '.' && s.[0] <> '_')
+  in
+  { sub_dirs = standard_dirs
+  ; data_only = Predicate_lang.empty
+  }
+
+let make ~sub_dirs ~ignored_sub_dirs ~data_only =
+  let sub_dirs =
+    let ignored_subdirs = Predicate_lang.union ignored_sub_dirs in
```
Comment (diml) [inline src/sub_dirs.ml]: shouldn't ignored_subdirs be merged with data_only rather than sub_dirs? To preserve backward compatibility

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/sub_dirs.ml (hunk 5 above)
Comment (rgrinberg) [reply src/sub_dirs.ml]: Good call.

**Review by ghost — APPROVED** [review]

**Pull request merged**